## Simulating Data

### msprime

Right now the functionality available is to simulate any number of epochs, and for each epoch generate a set of genotypes and phenotypes for the specified number of individuals. 
The script to run is 
```
python generate_simulated_phenogeno.py
```

### Input Flags

Input is provided as a series of flags when you run the program, as well as a file that lays out the specifics of each epoch that you want to simulate.

The potential input flags are:
```
-r, --repetitions 		number of simulated genotypes to run, corrisponding to
                    	the number of causal variants

-s, --samplesize 		number of genotypes that will be simulated for each
                    	run (Only required if input file is not provided; i.e. only one epoch is being simulated)

-w, --window 			window size to use as a spacer between the SNPs
                    	generated by the simulation. If the input is not a
                    	number, each SNP generated will be on its own
                    	chromosome (number = replicate)

-n, --num_inds 			number of individuals that will be used to create the
                    	genotypes and phenotypes (Only required if input file is not provided; i.e. only one epoch is being simulated)

-l, --sequence_length	length of the sequence that will be simulated; the
                    	only variant that will be used from this is the causal
                    	one

-b, --beta				string keyword that corrisponds to how we want to
                    	assign the beta value for the phenotype (ie constant
                    	means the beta value will be the same for all causal
                    	variants)

-o, --out 				name of the file to be saved, the files <out_name>.pheno.vcf 
						and <out_name>.phenotypes will be created

-f, --file 				the name of the file that contains the epochs that we
                    	will be simulating. The file format is described in
                    	the read_input_file function description.

-m, --mutrate 			rate of mutation to use for simulation.

-c, --recombrate 		rate of recombination to use for simulation.

-p, --phenotype          flag which causes the program to generate/save a phenotype file for each epoch

-a, --outofafrica        flag which must be passed if you want to use the "out of africa" model from the Gravel
                         2011 paper as the initial epoch. You also need to have the first entry in the input file
                         be a 'ooafrica' type of event

```

### Input File

Per epoch, you need:

#<type of event>	<Start time of epoch, unless it is an "ooafrica" event>	<population sizes for each population, seperated by commas>
Followed by (for "mrate" events):
```
<NxN migration matrix, where N is the number of populations>
```
or, for "mass" events; one per migration event at the given time:
```
<source population>	<destination population>	<proportion of source pop migrating> 
```
Here is an example of the input file:
```
#ooafrica       10,0,2
#mass   10000   4,7,7   
0       1       0.75
2       1       0.5
#mrate  100000  4,3,2
0       1e-7    1e-7
1e-7    0       1e-7
1e-7    1e-7    0

```

The type of event for the epoch is either "mass" (for a Mass Migration event), "mrate" for a change in the migration rate for that epoch, or "ooafrica" (as the first entry) to specify that the out-of-africa model should be used as the initial epoch.

For the "ooafrica" epoch, the population configuration must be three populations. The first population is the African population, second is European, and third is East Asian.

### Output

The output is two files per epoch: 
```
- a vcf file; <outname>.epoch<N>.pheno.vcf
- a phenotype file; <outname>.phenotypes (only if the -p flag is passed)
```

For the VCF file, each variant entry corrisponds to a single mutation generated in the course of a simulation  


Here is an example of using the script:

```
python generate_simulated_phenogeno.py -r 50 -l 1000 -w 1000 -f testing/test_inputfile.migration_events -m 2e-8 -c 1e-7 -o testing/test_massmig_input
```





