## Simulating Data

### msprime

Right now the functionality available is to simulate any number of epochs, and for each epoch generate a set of genotypes and phenotypes for the specified number of individuals. 
The script to run is 
```
python generate_simulated_phenogeno.py
```

Input is provided as a series of flags when you run the program, as well as a file that lays out the specifics of each epoch that you want to simulate.

The potential input flags are:
```
-r L, --repetitions L
                        number of simulated genotypes to run, corrisponding to
                        the number of causal variants
  -s SAMP_SIZE, --samplesize SAMP_SIZE
                        number of genotypes that will be simulated for each
                        run
  -w WINDOW, --window WINDOW
                        window size to use as a spacer between the SNPs
                        generated by the simulation. If the input is not a
                        number, each SNP generated will be on its own
                        cromosome (number = replicate)
  -n NUM_INDIVIDUALS, --num_inds NUM_INDIVIDUALS
                        number of individuals that will be used to create the
                        genotypes and phenotypes
  -l SEQ_LEN, --sequence_length SEQ_LEN
                        length of the sequence that will be simulated; the
                        only variant that will be used from this is the causal
                        one
  -b BETA_TYPE, --beta BETA_TYPE
                        string keyword that corrisponds to how we want to
                        assign the beta value for the phenotype (ie constant
                        means the beta value will be the same for all causal
                        variants)
  -o OUT_NAME, --out OUT_NAME
                        name of the file to be saved, the files
                        <out_name>.phenodist.png, <out_name>.pheno.vcf,
                        <out_name>.geno.vcf and <out_name>.phenotypes will be
                        created
  -f INPUT_FILE, --file INPUT_FILE
                        the name of the file that contains the epochs that we
                        will be simulating. The file format is described in
                        the read_input_file function description.
  -m MUTATION_RATE, --mutrate MUTATION_RATE
                        rate of mutation to use for simulation.
  -c RECOMBINATION_RATE, --recombrate RECOMBINATION_RATE
                        rate of recombination to use for simulation.


The output is two files per epoch: 
```
- a vcf file; <outname>.epoch<N>.pheno.vcf
- a phenotype file; <outname>.phenotypes
```

For the VCF file, each variant entry corrisponds to a single mutation generated in the course of a simulation  