{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pandas import *\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "from matplotlib import pyplot\n",
    "import itertools as it\n",
    "import matplotlib.backends.backend_pdf\n",
    "import math\n",
    "from matplotlib.pyplot import cm\n",
    "from dateutil import parser\n",
    "import scipy\n",
    "from scipy.stats import mstats\n",
    "import re\n",
    "import matplotlib.dates as mdates\n",
    "import datetime\n",
    "import msprime as ms\n",
    "import random\n",
    "from scipy.stats import norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_haplotypes(var_dict,samp_size,seq_len):\n",
    "    hap_list = ['' for x in range(samp_size)]\n",
    "    for pos in range(seq_len):\n",
    "        if(pos in var_dict.keys()):\n",
    "            for ind,al in enumerate(var_dict[pos]):\n",
    "                hap_list[ind] = ''.join([hap_list[ind],str(al)])\n",
    "#         else:\n",
    "#             for ind in range(samp_size):\n",
    "#                 hap_list[ind] = ''.join([hap_list[ind],'0'])\n",
    "    ind_haps_dict = {}\n",
    "    haps_used = []\n",
    "    x = int(samp_size/2)\n",
    "    for n in range(x):\n",
    "        curr_haps = []\n",
    "        while (len(curr_haps) != 2):\n",
    "            temp_seq_num = random.randint(0,(samp_size-1))\n",
    "            if(temp_seq_num not in haps_used):\n",
    "                curr_haps.append(hap_list[temp_seq_num])\n",
    "                haps_used.append(temp_seq_num)\n",
    "        ind_haps_dict[n] = curr_haps\n",
    "    return ind_haps_dict\n",
    "\n",
    "def generate_betas(num_inds,dist_type='normal'):\n",
    "    if(dist_type == 'normal'):\n",
    "        dist = np.random.normal(0,0.1,num_inds)\n",
    "    return dist\n",
    "\n",
    "\n",
    "def assign_genotype_index(samp_size,num_inds):\n",
    "    ind_haps_dict = {}\n",
    "    haps_used = []\n",
    "    x = num_inds\n",
    "#     x = int(samp_size/2)\n",
    "    for n in range(x):\n",
    "        curr_haps = []\n",
    "        while (len(curr_haps) != 2):\n",
    "            temp_seq_num = random.randint(0,(samp_size-1))\n",
    "            if(temp_seq_num not in haps_used and temp_seq_num not in curr_haps):\n",
    "                curr_haps.append(temp_seq_num)\n",
    "                haps_used.append(temp_seq_num)\n",
    "        ind_haps_dict[n] = curr_haps\n",
    "    return ind_haps_dict\n",
    "\n",
    "\n",
    "\n",
    "def estimate_pheno(genotypes,beta):\n",
    "#     causal_pos = 0 #int(len(genotype[0])/2)\n",
    "    full_phenovals = []\n",
    "    for g in genotypes:\n",
    "        try:\n",
    "            c_al1 = float(g[0])\n",
    "            c_al2 = float(g[1])\n",
    "            full_phenovals.append((c_al1 + c_al2)*beta)\n",
    "        except:\n",
    "            print(\"Error!\",g)\n",
    "    return sum(full_phenovals)\n",
    "\n",
    "def plot_phenodist(pheno_dict):\n",
    "    pheno_df = DataFrame(pheno_dict.values(),index=pheno_dict.keys(),columns=['Pheno_value'])\n",
    "\n",
    "    mu, std = norm.fit(pheno_df['Pheno_value'])\n",
    "\n",
    "    pyplot.hist(pheno_df['Pheno_value'])\n",
    "\n",
    "    xmin, xmax = pyplot.xlim()# min(pheno_df['Pheno_value']),max(pheno_df['Pheno_value'])\n",
    "    x = np.linspace(xmin, xmax, 100)\n",
    "#     p = (norm.pdf(x, mu, std))*(num_inds)\n",
    "    p = (norm.pdf(x, mu, std))*(len(pheno_df)*2)\n",
    "    pyplot.plot(x, p, 'k', linewidth=2)\n",
    "    title = \"Fit results: mu = %.2f,  std = %.2f\" % (mu, std)\n",
    "    pyplot.title(title)\n",
    "    pyplot.ylabel('Number of Individuals in Bin')\n",
    "    pyplot.xlabel('Phenotype Value')\n",
    "#     pyplot.savefig('/local3/jake/admix_simul/pheno_plot_test.png')\n",
    "    pyplot.show()\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def write_phenofile(outname,pheno_dict,popid_dict=None):\n",
    "    pheno_file = open('{0}.phenotypes'.format(outname),'w')\n",
    "    if(popid_dict == None):\n",
    "        new_ids = []\n",
    "        for i in range(len(pheno_dict.items())):\n",
    "            new_ids.append(''.join(['ID',str(i)]))\n",
    "    else:\n",
    "        new_ids = []\n",
    "        for i,pop in popid_dict.items():\n",
    "            new_ids.append(''.join(['POP',str(pop),'ID',str(i),]))\n",
    "    for ind,p in pheno_dict.items():\n",
    "        pheno_file.write('{0}\\t{1}\\n'.format(new_ids[ind],p))\n",
    "    pheno_file.close()\n",
    "\n",
    "def run_pheno_simulation(samp_size,seq_len,L,num_individuals,causal_var_id=1):\n",
    "    genotype_index_byinds = assign_genotype_index(samp_size,num_individuals) #randomly take the <samp_size> number of genomes simulated, and randomly assign to each individual 2 of them\n",
    "#     causalgenotypes_byind = {x:[] for x in range(num_individuals)} #Each repetition, add the genotype from that rep (as a tuple) to this dictionary. Then we can just iterate through the list of values to generate the final phenotype\n",
    "    causalgenotypes_byrep = {x:[] for x in range(L)}\n",
    "    causalpositions_byrep = {x:0 for x in range(L)}\n",
    "    fullgenotypes_byrep = {x:[] for x in range(L)}\n",
    "    rep = 0\n",
    "    while (rep < L):\n",
    "        tree_sequence = ms.simulate(sample_size=samp_size, Ne=1e4, length=seq_len, recombination_rate=2e-8,mutation_rate=2e-8) \n",
    "        curr_causal_var = []\n",
    "        curr_causal_pos = 0\n",
    "        curr_full_vars_posgeno_dict = {}\n",
    "        for variant in tree_sequence.variants():\n",
    "            curr_full_vars_posgeno_dict[round(variant.site.position)] = list(variant.genotypes)\n",
    "            if(variant.site.id == causal_var_id):\n",
    "                curr_causal_var = list(variant.genotypes)\n",
    "                curr_causal_pos = round(variant.site.position)\n",
    "\n",
    "        if(len(curr_causal_var) != samp_size):\n",
    "            print('no causal variants in rep {0}, redoing'.format(rep))\n",
    "            rep -= 1\n",
    "            continue\n",
    "        causalpositions_byrep[rep] = curr_causal_pos\n",
    "        curr_rep_genotypes = []\n",
    "        for indiv,index in genotype_index_byinds.items():\n",
    "            try:\n",
    "                causalgenotypes_byind[indiv].append((curr_causal_var[index[0]],curr_causal_var[index[1]]))\n",
    "                curr_rep_genotypes.append((curr_causal_var[index[0]],curr_causal_var[index[1]]))\n",
    "            except:\n",
    "                print(indiv,index)\n",
    "        causalgenotypes_byrep[rep] = [curr_causal_pos,curr_rep_genotypes]\n",
    "        \n",
    "        curr_rep_fullgenos = {}\n",
    "        for pos,geno in curr_full_vars_posgeno_dict.items():\n",
    "            temp_fullgeno = []\n",
    "            for indiv,index in genotype_index_byinds.items():\n",
    "                try:\n",
    "                    temp_fullgeno.append((geno[index[0]],geno[index[1]]))\n",
    "                except:\n",
    "                    print(indiv,index)\n",
    "            curr_rep_fullgenos[pos] = temp_fullgeno\n",
    "        fullgenotypes_byrep[rep] = curr_rep_fullgenos\n",
    "        rep += 1\n",
    "\n",
    "    phenotypes_byinds = {x:0 for x in range(num_individuals)}\n",
    "    \n",
    "    beta_list = generate_betas(num_inds=num_individuals)\n",
    "    for i in range(num_individuals):\n",
    "        curr_beta = beta_list[i]\n",
    "        phenotypes_byinds[i] = estimate_pheno(causalgenotypes_byind[i],curr_beta)\n",
    "        \n",
    "    plot_phenodist(phenotypes_byinds)\n",
    "    write_phenofile('test_phenos',phenotypes_byinds)\n",
    "    \n",
    "    return causalpositions_byrep,fullgenotypes_byrep\n",
    "    \n",
    "   \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_geno_simulation(samp_size,seq_len,L,num_individuals,causal_var_id=1):\n",
    "    genotype_index_byinds = assign_genotype_index(samp_size,num_individuals) #randomly take the <samp_size> number of genomes simulated, and randomly assign to each individual 2 of them\n",
    "    causalgenotypes_byind = {x:[] for x in range(num_individuals)} #Each repetition, add the genotype from that rep (as a tuple) to this dictionary. Then we can just iterate through the list of values to generate the final phenotype\n",
    "    causalgenotypes_byrep = {x:[] for x in range(L)}\n",
    "    rep = 0\n",
    "    while (rep < L):\n",
    "        tree_sequence = ms.simulate(sample_size=samp_size, Ne=1e4, length=seq_len, recombination_rate=2e-8,mutation_rate=2e-8) \n",
    "        num_vars = 0\n",
    "        for variant in tree_sequence.variants():\n",
    "            num_vars += 1\n",
    "        causal_var_id = temp_seq_num = random.randint(0,(num_vars-1))\n",
    "        curr_causal_var = []\n",
    "        curr_causal_pos = 0\n",
    "        for variant in tree_sequence.variants():\n",
    "            if(variant.site.id == causal_var_id):\n",
    "                curr_causal_var = list(variant.genotypes)\n",
    "                curr_causal_pos = round(variant.site.position)\n",
    "\n",
    "        if(len(curr_causal_var) != samp_size):\n",
    "            print('no causal variants in rep {0}, redoing'.format(rep))\n",
    "            rep -= 1\n",
    "            continue\n",
    "        curr_rep_genotypes = []\n",
    "        for indiv,index in genotype_index_byinds.items():\n",
    "            try:\n",
    "                curr_rep_genotypes.append((curr_causal_var[index[0]],curr_causal_var[index[1]]))\n",
    "#                 causalgenotypes_byind[indiv].append((curr_causal_var[index[0]],curr_causal_var[index[1]]))\n",
    "            except:\n",
    "                print(indiv,index)\n",
    "        causalgenotypes_byrep[rep] = [curr_causal_pos,curr_rep_genotypes]\n",
    "        rep += 1\n",
    "    return causalgenotypes_byrep\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_genovcf(full_geno_dict,causal_pos_dict,outname,seq_len,window_spacer=1000):\n",
    "#     full_len = len(full_geno_dict)*(seq_len + (window_spacer*2))\n",
    "    full_len = len(full_geno_dict)*window_spacer\n",
    "    header_string = '##fileformat=VCFv4.2 \\n##source=tskit 0.2.2 \\n##FILTER=<ID=PASS,Description=\"All filters passed\"> \\n##INFO=<ID=CS,Number=0,Type=Flag,Description=\"SNP Causal to Phenotype\"> \\n##contig=<ID=1,length={0}>\\n##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\"> \\n'.format(full_len)\n",
    "    vcf_file = open('/local3/jake/admix_simul/testing/{0}.vcf'.format(outname),'w')\n",
    "    vcf_file.write(header_string)\n",
    "    vcf_file.write('#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\t')\n",
    "    new_ids = []\n",
    "    for i in range(len(full_geno_dict[0][causal_pos_dict[0]])):\n",
    "        curr_nid = ''.join(['ID',str(i)])\n",
    "        new_ids.append(curr_nid)\n",
    "        vcf_file.write('{0}\\t'.format(curr_nid))\n",
    "    vcf_file.write('\\n')\n",
    "    for rep,geno_dict in full_geno_dict.items():\n",
    "        is_rep_causalsnp = False\n",
    "        for pos,genos in geno_dict.items():\n",
    "            try:\n",
    "                curr_pos = (int(rep)*window_spacer)+pos\n",
    "                if(pos == causal_pos_dict[rep]):\n",
    "                    is_rep_causalsnp = True\n",
    "                    vcf_file.write('1\\t{0}\\t.\\tA\\tG\\t.\\tPASS\\tCS\\tGT\\t'.format(curr_pos))\n",
    "                else:\n",
    "                    vcf_file.write('1\\t{0}\\t.\\tA\\tG\\t.\\tPASS\\t.\\tGT\\t'.format(curr_pos))\n",
    "                for g in genos:\n",
    "                    vcf_file.write('{0}|{1}\\t'.format(g[0],g[1]))\n",
    "                vcf_file.write('\\n')\n",
    "            except:\n",
    "                print(rep,genos)\n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_input_migrationmatrix(filename,num_epochs):\n",
    "    mfile = open(filename,'r')\n",
    "    migmats_byepoch = [[] for x in range(num_epochs)]\n",
    "    for line in mfile:\n",
    "        if(line[0] == '#'):\n",
    "            curr_migmat = int(line[1])\n",
    "        else:\n",
    "            curr_mm = np.array([float(f) for f in line.split('\\n')[0].split('\\t')])\n",
    "            migmats_byepoch[curr_migmat].append(curr_mm)\n",
    "    mfile.close()\n",
    "    return migmats_byepoch\n",
    "def read_input_file(filename):\n",
    "    input_df = read_csv('{0}.epochs'.format(filename),sep='\\t')\n",
    "    num_theta = len(input_df)\n",
    "    theta_true = []\n",
    "    sample_scheme_byepoch = []\n",
    "    migmat_fromfile = None\n",
    "    try:\n",
    "        mm_file = '{0}.migration_matrix'.format(filename)\n",
    "        migmat_fromfile = get_input_migrationmatrix(filename,num_theta)\n",
    "    except:\n",
    "        pass\n",
    "    # Set the demographic model parameters (theta)...\n",
    "    for epoch,t in input_df.iterrows():\n",
    "        Np = len(t['pop_scheme'].split(','))               # Num pops\n",
    "        pop_sizes = [int(x) for x in t['pop_scheme'].split(',')]    # Num of samples per population in the observed data\n",
    "        sample_scheme = [int(x) for x in t['pop_scheme'].split(',')]   # Num samples drawn from each pop for likelihood\n",
    "\n",
    "        # pop sizes\n",
    "        N_base = 1e4    # Number of diploids in the population\n",
    "        Ne0 = np.ones(Np)*N_base\n",
    "        \n",
    "        if(migmat_fromfile is None):\n",
    "            # Set migration rates\n",
    "            M_base = float(t['migration_rate'])\n",
    "            m = np.zeros((Np,Np))\n",
    "            for i in range(Np):\n",
    "                for j in range(Np):\n",
    "                    if j!=i:\n",
    "                        m[i,j] = M_base/(2*(Np-1))\n",
    "            theta_true.append([float(t['time']),[Ne0,m]])\n",
    "        else:\n",
    "            theta_true.append([float(t['time']),migmat_fromfile[epoch]])\n",
    "        sample_scheme_byepoch.append(pop_sizes)\n",
    "\n",
    "    return theta_true,sample_scheme_byepoch\n",
    "\n",
    "\n",
    "#Reads in a file that has the epochs that we want to simulate. The file needs to be in a tab-seperated file, with a line per epoch, each one followed by a migration matrix outlining the migration values for each population pair\n",
    "#Each line that indicates an epoch should start with a '#' symbol, followed immediately by the time of the start of the epoch, then a tab, then the population sizes seperated by commas (for example: 4,4,5 for 3 populations of that number of individuals)\n",
    "#Following the line with the time/population sizes, the migration matrix which is of size NxN, where N=number of populations. The diagonal of the matrix should be zeros.\n",
    "#   filename (str): name of the input file to read\n",
    "#returns:\n",
    "#   theta_true: description can be found in the run_\n",
    "def read_input_file_full(filename):\n",
    "    infile = open(filename,'r')\n",
    "    theta_true = []\n",
    "    sample_scheme_byepoch = []\n",
    "    curr_epoch = -1\n",
    "    migmats_byepoch = {}\n",
    "    times_byepoch = []\n",
    "    for line in infile:\n",
    "        if(line[0] == '#'):\n",
    "            curr_epoch += 1\n",
    "            curr_time = line.split('\\n')[0][1:].split('\\t')[0]\n",
    "            times_byepoch.append(float(curr_time))\n",
    "            curr_popscheme = line.split('\\n')[0][1:].split('\\t')[1]\n",
    "            pop_sizes = [int(x) for x in curr_popscheme.split(',')]    # Num of samples per population in the observed data\n",
    "            sample_scheme_byepoch.append(pop_sizes)\n",
    "        else:\n",
    "            curr_mm = [float(f) for f in line.split('\\n')[0].split('\\t')]\n",
    "            if(curr_epoch not in migmats_byepoch.keys()):\n",
    "                migmats_byepoch[curr_epoch] = []\n",
    "            migmats_byepoch[curr_epoch].append(curr_mm)\n",
    "    # pop sizes\n",
    "    N_base = 1e4    # Number of diploids in the population\n",
    "    \n",
    "    for e in range(len(sample_scheme_byepoch)):\n",
    "        Ne0 = np.ones(len(sample_scheme_byepoch[e]))*N_base\n",
    "        theta_true.append([times_byepoch[e],[Ne0,np.array(migmats_byepoch[e])]])\n",
    "\n",
    "    return theta_true,sample_scheme_byepoch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_phenovcf(pheno_dict,outname,seq_len,pop_dict=None):\n",
    "    header_string = '##fileformat=VCFv4.2 \\n##source=tskit 0.2.2 \\n##FILTER=<ID=PASS,Description=\"All filters passed\"> \\n##contig=<ID=1,length=1000>\\n##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\"> \\n'\n",
    "    vcf_file = open('/local3/jake/admix_simul/testing/{0}.vcf'.format(outname),'w')\n",
    "    vcf_file.write(header_string)\n",
    "    vcf_file.write('#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\t')\n",
    "    if(pop_dict == None):\n",
    "        new_ids = []\n",
    "        for i in range(len(geno_dict[0][1])):\n",
    "            curr_nid = ''.join(['ID',str(i)])\n",
    "            new_ids.append(curr_nid)\n",
    "            vcf_file.write('{0}\\t'.format(curr_nid))\n",
    "        vcf_file.write('\\n')\n",
    "    else:\n",
    "        new_ids = []\n",
    "        for i,pop in popid_dict.items():\n",
    "            curr_nid = ''.join(['POP',str(pop),'ID',str(i),])\n",
    "            new_ids.append(curr_nid)\n",
    "            vcf_file.write('{0}\\t'.format(curr_nid))\n",
    "        vcf_file.write('\\n')\n",
    "    for rep,genos in geno_dict.items():\n",
    "        curr_pos = (int(rep)*seq_len)+genos[0]\n",
    "        vcf_file.write('1\\t{0}\\t.\\tA\\tG\\t.\\tPASS\\t.\\tGT\\t'.format(curr_pos))\n",
    "        for g in genos[1]:\n",
    "            vcf_file.write('{0}|{1}\\t'.format(g[0],g[1]))\n",
    "        vcf_file.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_indtopop_dict(genoind_bypops):\n",
    "    indpop_dict = {}\n",
    "    prev_pn = 0\n",
    "    for num,p in enumerate(genoind_bypops.items()):\n",
    "        for i,g in enumerate(p[1]):\n",
    "            indpop_dict[(prev_pn+i)] = p[0]\n",
    "        prev_pn += len(p[1])\n",
    "    return indpop_dict\n",
    "\n",
    "def get_ind_genoindex_multipop(genoind_bypops):\n",
    "    ind_genoindex_dict = {}\n",
    "    prev_pn = 0\n",
    "    for num,p in enumerate(genoind_bypops.items()):\n",
    "        for i,g in enumerate(p[1]):\n",
    "            ind_genoindex_dict[(prev_pn+i)] = g\n",
    "        prev_pn += len(p[1])\n",
    "    return ind_genoindex_dict\n",
    "\n",
    "def assign_pop_inds(pop_sizes):\n",
    "    pop_assign = []  # row index = pop index. value = list of indices of samples assigned to each pop.\n",
    "    j = 0\n",
    "    for i in range(len(pop_sizes)):\n",
    "        Ni = pop_sizes[i]\n",
    "        indices = list(range(j,j+Ni))\n",
    "        pop_assign.append(indices)\n",
    "        j+=Ni\n",
    "    return pop_assign\n",
    "def assign_popdict(pop_scheme):\n",
    "    full_sampsize = sum(pop_scheme)\n",
    "    popindex_byind = {x:0 for x in range(full_sampsize)}\n",
    "    for pop,size in enumerate(pop_scheme):\n",
    "        for x in range(size):\n",
    "            popindex_byind[(x + (size*pop))] = pop\n",
    "    return popindex_byind\n",
    "\n",
    "#Take the genotypes that are simulated, and assign two of the indexes to each individual.\n",
    "#samp_sizes: number of individuals that you want to sample, an array of length P\n",
    "#pop_sizes: Total number of genotypes in each population that you are simulating, an array of length P. Each entry must be at least double the entry in samp_sizes\n",
    "def assign_genotype_index_multipop(samp_sizes,pop_sizes):\n",
    "    for i,s in enumerate(samp_sizes):\n",
    "        if(pop_sizes[i] < (2*s)):\n",
    "            print('number of genotypes in population {0} is not at least double the number of individuals you are simulating! {1} < {2}'.format(i,pop_sizes[i],(s*2)))\n",
    "            return -1\n",
    "    ind_haps_dict_bypop = {x:[] for x in range(len(pop_sizes))}\n",
    "    x = sum(pop_sizes)\n",
    "    prev_indstart = 0\n",
    "    for n,p in enumerate(pop_sizes):\n",
    "        num_inds_tosample = samp_sizes[n]\n",
    "        curr_popinds = []\n",
    "        for x in range(p):\n",
    "            curr_popinds.append(x + prev_indstart)\n",
    "        \n",
    "        print(n,p,num_inds_tosample,curr_popinds)\n",
    "        popinds_used = []\n",
    "        temp_popinds = []\n",
    "        for x in range(num_inds_tosample):\n",
    "            curr_haps = []\n",
    "            while (len(curr_haps) != 2):\n",
    "                temp_seq_num = random.randint(min(curr_popinds),max(curr_popinds))\n",
    "                if(temp_seq_num not in popinds_used and temp_seq_num not in curr_haps):\n",
    "                    curr_haps.append(temp_seq_num)\n",
    "                    popinds_used.append(temp_seq_num)\n",
    "            temp_popinds.append(curr_haps)\n",
    "        ind_haps_dict_bypop[n] = temp_popinds\n",
    "        prev_indstart += p\n",
    "    return ind_haps_dict_bypop\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_theta,test_sampsch = read_input_file_full_v1('/local3/jake/admix_simul/testing/test_inputfile.migration_events')\n",
    "# test_theta,test_sampsch = read_input_file_full('/local3/jake/admix_simul/testing/test_inputfile.migration_matrix')\n",
    "test_theta,test_sampsch = read_input_file_full_v1('/local3/jake/admix_simul/testing/single_massmig_event/input_file.migr_events.v2')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['ooafrica', [100, 0, 0]],\n",
       " ['mass', 120.0, [array([12300., 12300., 12300.]), array([[1. , 0. , 0.5]])]]]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_theta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting epoch 0 replicates\n",
      "0 8 4 [0, 1, 2, 3, 4, 5, 6, 7]\n",
      "1 8 4 [8, 9, 10, 11, 12, 13, 14, 15]\n",
      "2 8 4 [16, 17, 18, 19, 20, 21, 22, 23]\n",
      "there is only one epoch, starting now using theta [['mrate', 0.0, [array([10000., 10000., 10000.]), array([[0.    , 0.0001, 0.01  ],\n",
      "       [0.0001, 0.    , 0.001 ],\n",
      "       [0.01  , 0.001 , 0.    ]])]]]\n",
      "replicate trees created; iterating through genotypes now\n",
      "finished iterating through trees, starting the phenotype calculations\n",
      "Starting epoch 1 replicates\n",
      "0 8 4 [0, 1, 2, 3, 4, 5, 6, 7]\n",
      "1 14 7 [8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21]\n",
      "2 14 7 [22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35]\n",
      "Epoch 1 has no migration, skipping this epoch due to memory constraints\n",
      "Starting epoch 2 replicates\n",
      "0 8 4 [0, 1, 2, 3, 4, 5, 6, 7]\n",
      "1 14 7 [8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21]\n",
      "2 14 7 [22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35]\n",
      "Epoch 2 has no migration, skipping this epoch due to memory constraints\n",
      "Starting epoch 3 replicates\n",
      "0 8 4 [0, 1, 2, 3, 4, 5, 6, 7]\n",
      "1 6 3 [8, 9, 10, 11, 12, 13]\n",
      "2 4 2 [14, 15, 16, 17]\n",
      "replicate trees created; iterating through genotypes now\n",
      "finished iterating through trees, starting the phenotype calculations\n"
     ]
    }
   ],
   "source": [
    "mu=1e-7\n",
    "r=2e-8  \n",
    "cpos_pheno_byinds,pgenos_byinds = run_pheno_simulation_multipops(test_theta,1000,25,test_sampsch,r,mu,'/local3/jake/admix_simul/testing/test_output_massmig','normal')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def thetas_toskip(theta):\n",
    "    bad_theta_nums = []\n",
    "    for epoch in range(len(theta)):\n",
    "        curr_theta = theta[:epoch+1]\n",
    "        if(curr_theta[-1][0] == 'mass'):\n",
    "            bad_theta_nums.append(epoch)\n",
    "        elif(set([all(x == 0) for x in curr_theta[-1][2][1]]) == {True}):\n",
    "            bad_theta_nums.append(epoch)\n",
    "    return bad_theta_nums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mrate'"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# thetas_toskip(test_theta)\n",
    "test_theta[-1][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#theta: the list of epochs that we want to simulate. Each entry is equal to one epoch; each entry has:\n",
    "#  0: Starting time of the epoch\n",
    "#  1: List with two items: Total Population size at start of epoch, and migration matrix\n",
    "#    Migration Matrix: NxN (where N = Number of Populations, Np), Rate of migration between each population\n",
    "#sample_nums: Array of population sizes, of length Np; the number of samples drawn from the population\n",
    "#L: Length of the sequence we will be simulating\n",
    "#r: Recombination rate, constant accross all populations\n",
    "#mu: Mutation rate, constant accross all populations\n",
    "#R: Number of replicates; msprime will run R simulations and return an iterator over all trees created\n",
    "#Updated: Return list of ts_replicates; each entry will be the simulations for each epoch\n",
    "def run_msprime_tskit(theta,sample_nums,L,r,mu,R):\n",
    "    ts_replicates = None\n",
    "    \n",
    "    Np = len(sample_nums)\n",
    "#     print('samples to be drawn for populations: {0}'.format(sample_nums))\n",
    "    init_pop_configs = [ms.PopulationConfiguration(sample_size=sample_nums[i], initial_size=theta[0][2][0][i]) for i in range(Np)]\n",
    "\n",
    "    if(theta[0][0] == 'mrate'):\n",
    "        init_mig = theta[0][2][1]\n",
    "    elif(theta[0][0] == 'mass'):\n",
    "        init_mig = None\n",
    "        init_demoevents = []\n",
    "        for m in theta[0][2][1]:\n",
    "            init_demoevents.append(ms.MassMigration(time=0, source=m[0], destination=m[1], proportion=m[2]))\n",
    "    K = len(theta)  # K = number of epochs\n",
    "    if K > 1:\n",
    "        # There is more than one epoch, so must set the non-initial epochs as demographic events\n",
    "        demo_events = []\n",
    "        for k in range(1,K):\n",
    "            theta_type,t_k,theta_k = theta[k]\n",
    "            Ne = theta_k[0]\n",
    "            if(theta_type == 'mrate'):\n",
    "                mig = theta_k[1]\n",
    "                for i in range(Np):\n",
    "                    # Set the Ne\n",
    "                    demo_events.append(ms.PopulationParametersChange(population=i,time=t_k,initial_size=Ne[i]))\n",
    "\n",
    "                    # Set the migration rates\n",
    "                    for j in range(Np):\n",
    "                        if j!=i:\n",
    "                            demo_events.append(ms.MigrationRateChange(time=t_k,rate=mig[i,j],matrix_index=tuple([i,j])))\n",
    "            elif(theta_type == 'mass'):\n",
    "                massmig = theta_k[1]\n",
    "                for m in massmig:\n",
    "                    demo_events.append(ms.MassMigration(time=t_k, source=m[0], destination=m[1], proportion=m[2]))\n",
    "            else:\n",
    "                print('Epoch events should only be \"mass\" or \"mrate\"!')\n",
    "                return -1\n",
    "#         print(init_pop_configs)\n",
    "        ts_replicates = ms.simulate(\n",
    "            length=L,\n",
    "            recombination_rate=r,\n",
    "            population_configurations=init_pop_configs,\n",
    "            migration_matrix = init_mig,\n",
    "            demographic_events=demo_events,\n",
    "            num_replicates=R,\n",
    "            mutation_rate = mu\n",
    "        )\n",
    "    else:\n",
    "        print('there is only one epoch, starting now using theta {0}'.format(theta))\n",
    "        # There is only the initial epoch. \n",
    "        if(init_mig is None): #If the first epoch event is a mass migration event, then there's no migration matrix to provide\n",
    "            ts_replicates = ms.simulate(\n",
    "            length=L,\n",
    "            recombination_rate=r,\n",
    "            population_configurations=init_pop_configs,\n",
    "            demographic_events = init_demoevents,\n",
    "            num_replicates=R,\n",
    "            mutation_rate = mu)\n",
    "        else:\n",
    "            ts_replicates = ms.simulate(\n",
    "                length=L,\n",
    "                recombination_rate=r,\n",
    "                population_configurations=init_pop_configs,\n",
    "                migration_matrix = init_mig,\n",
    "                num_replicates=R,\n",
    "                mutation_rate = mu\n",
    "            )\n",
    "    return ts_replicates\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fixed model parameters...\n",
    "mu=1e-7                                  # mutation rate per site per generation per lineage\n",
    "r=2e-8                                   # recombination rate per site per generation per lineage\n",
    "Lobs = int(2e4)                          # The total number of sites in the observed sequence data\n",
    "R=20                                     # Num of replicate ARGs to run per model\n",
    "L=Lobs                                   # size of each ARG window \n",
    "windows = [[0,L]]                        # List of the windows to use\n",
    "draws_max = 1                            # Maximum number of sample draws to use\n",
    "tol = 0.5                                # Tolerance in fraction of mutational patterns that don't match the ARG tree among usable sites\n",
    "order='natural'                             # The sample label order to use. 'natural'= do nothing. 'WPDO'= put samples in Within-Population Derived-allele Order\n",
    "\n",
    "# Set the demographic model parameters (theta)...\n",
    "Np = 3                                  # Num pops\n",
    "pop_sizes = [4 for i in range(Np)]      # Num of samples per population in the observed data\n",
    "sample_scheme = [4 for i in range(Np)]  # Num samples drawn from each pop for likelihood\n",
    "\n",
    "# pop sizes\n",
    "N_base = 1e4    # Number of diploids in the population\n",
    "Ne0 = np.ones(Np)*N_base\n",
    "\n",
    "# Set migration rates\n",
    "M_base = 1e-4\n",
    "m0 = np.zeros((Np,Np))\n",
    "m1 = m0.copy()\n",
    "for i in range(Np):\n",
    "    for j in range(Np):\n",
    "        if j!=i:\n",
    "            m1[i,j] = M_base/(2*(Np-1))\n",
    "            \n",
    "phi0 = [Ne0,m0]\n",
    "phi1 = [Ne0,m1]\n",
    "t1 = 1e4\n",
    "theta_true=[[0.,phi0],[t1,phi1]]\n",
    "\n",
    "pops = [i for i in range(Np)]\n",
    "n = np.sum(sample_scheme)\n",
    "N = np.sum(pop_sizes)\n",
    "pop_assign = []  # row index = pop index. value = list of indices of samples assigned to each pop.\n",
    "j = 0\n",
    "for i in range(Np):\n",
    "    Ni = pop_sizes[i]\n",
    "    indices = list(range(j,j+Ni))\n",
    "    pop_assign.append(indices)\n",
    "    j+=Ni\n",
    "\n",
    "# Generate the fake seq data from the true model, S\n",
    "# S = simulate_sequence_data(theta_true,pop_sizes,Lobs,r,mu)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting epoch 1 replicates\n",
      "Starting epoch 2 replicates\n"
     ]
    }
   ],
   "source": [
    "# test_tsreps = run_msprime_tskit(theta_true,sample_scheme,1000,r,mu,R)\n",
    "# run_pheno_simulation_multipops(theta_true,1000,50,3,4)\n",
    "theta = theta_true\n",
    "seq_len = 1000\n",
    "reps = 50\n",
    "r = 2e-8 #recombination rate\n",
    "mu = 2e-8 #mutation rate\n",
    "num_pops = 3\n",
    "num_individuals = 4\n",
    "\n",
    "pop_sizes = [(num_individuals*2) for i in range(num_pops)] # Num individuals simulated, equal to twice the number of individuals, since each person is diploid\n",
    "sample_scheme = [num_individuals for i in range(num_pops)] # Num samples drawn from each pop for likelihood\n",
    "full_sampsize = sum(sample_scheme)\n",
    "\n",
    "genotype_index_bypops = assign_genotype_index_multipop(sample_scheme,pop_sizes) #randomly take the <samp_size> number of genomes simulated, and randomly assign to each individual 2 of them\n",
    "genotype_index_byinds = assign_genotype_index(sum(pop_sizes),full_sampsize)\n",
    "causalgenotypes_byrep = {x:[] for x in range(reps)}\n",
    "causalpositions_byrep = {x:0 for x in range(reps)}\n",
    "fullgenotypes_byrep = {x:[] for x in range(reps)}\n",
    "\n",
    "\n",
    "phenos_byepoch = []\n",
    "genos_byepoch = []\n",
    "causalpos_byepoch = []\n",
    "popindex_byind = {x:0 for x in range(full_sampsize)}\n",
    "for pop,size in enumerate(sample_scheme):\n",
    "    for x in range(size):\n",
    "        popindex_byind[(x + (size*pop))] = pop\n",
    "\n",
    "for epoch in range(len(theta)):\n",
    "    print('Starting epoch {0} replicates'.format(epoch+1))\n",
    "#     print(epoch,len(theta[:epoch+1]))\n",
    "    curr_theta = theta[:epoch+1]\n",
    "    new_sampscheme = [8 for x in range(num_pops)]\n",
    "    tsreps = run_msprime_tskit_updated(curr_theta,pop_sizes,seq_len,r,mu,reps)\n",
    "    for rep,tree_sequence in enumerate(tsreps): \n",
    "        num_vars = 0\n",
    "        for variant in tree_sequence.variants():\n",
    "            num_vars += 1\n",
    "        causal_var_id = random.randint(0,(num_vars-1))\n",
    "        curr_causal_var = []\n",
    "        curr_causal_pos = 0\n",
    "        curr_full_vars_posgeno_dict = {}\n",
    "        for variant in tree_sequence.variants():\n",
    "#             print(len(list(variant.genotypes)))\n",
    "            curr_full_vars_posgeno_dict[round(variant.site.position)] = list(variant.genotypes)\n",
    "            if(variant.site.id == causal_var_id):\n",
    "                curr_causal_var = list(variant.genotypes)\n",
    "                curr_causal_pos = round(variant.site.position)\n",
    "\n",
    "        if(len(curr_causal_var) != (full_sampsize*2)):\n",
    "            print('not enough causal variants in rep {0}'.format(rep))\n",
    "            curr_causal_var = [0 for x in range((full_sampsize*2))]\n",
    "            curr_full_vars_posgeno_dict[0] = [0 for x in range((full_sampsize*2))]\n",
    "        causalpositions_byrep[rep] = curr_causal_pos\n",
    "\n",
    "        curr_rep_genotypes = []\n",
    "        for indiv,index in genotype_index_byinds.items():\n",
    "            try:\n",
    "                curr_rep_genotypes.append((curr_causal_var[index[0]],curr_causal_var[index[1]]))\n",
    "            except:\n",
    "                print(indiv,index,len(curr_causal_var))\n",
    "        causalgenotypes_byrep[rep] = [curr_causal_pos,curr_rep_genotypes]\n",
    "\n",
    "        curr_rep_fullgenos = {}\n",
    "        for pos,geno in curr_full_vars_posgeno_dict.items():\n",
    "            temp_fullgeno = []\n",
    "            for indiv,index in genotype_index_byinds.items():\n",
    "                try:\n",
    "                    temp_fullgeno.append((geno[index[0]],geno[index[1]]))\n",
    "                except:\n",
    "                    print(rep,indiv,index,len(geno))\n",
    "            curr_rep_fullgenos[pos] = temp_fullgeno\n",
    "        fullgenotypes_byrep[rep] = curr_rep_fullgenos\n",
    "        \n",
    "    causalgenotypes_byind = {x:[] for x in range(full_sampsize)}\n",
    "    for pos,genos in causalgenotypes_byrep.items():\n",
    "        for num,g in enumerate(genos[1]):\n",
    "            causalgenotypes_byind[num].append(g)\n",
    "    beta = 'normal'\n",
    "    phenotypes_byinds = {x:0 for x in range(full_sampsize)}\n",
    "    beta_list = generate_betas(num_inds=full_sampsize,dist_type=beta)\n",
    "    for i in range(full_sampsize):\n",
    "        curr_beta = beta_list[i]\n",
    "        phenotypes_byinds[i] = estimate_pheno(causalgenotypes_byind[i],curr_beta)\n",
    "    phenos_byepoch.append(phenotypes_byinds)\n",
    "    genos_byepoch.append(fullgenotypes_byrep)\n",
    "    causalpos_byepoch.append(causalpositions_byrep)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_genovcf(full_geno_dict,causal_pos_dict,outname,num_inds,seq_len,window_spacer=1000,popid_dict=None):\n",
    "    use_diff_chrms = False\n",
    "    if(window_spacer == 'chrm'):\n",
    "        window_spacer = 0\n",
    "        use_diff_chrms = True\n",
    "    full_len = len(full_geno_dict)*(window_spacer+seq_len)\n",
    "    header_string = '##fileformat=VCFv4.2 \\n##source=tskit 0.2.2 \\n##FILTER=<ID=PASS,Description=\"All filters passed\"> \\n##INFO=<ID=CS,Number=0,Type=Flag,Description=\"SNP Causal to Phenotype\"> \\n##contig=<ID=1,length={0}>\\n##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\"> \\n'.format(full_len)\n",
    "    vcf_file = open('{0}.vcf'.format(outname),'w')\n",
    "    vcf_file.write(header_string)\n",
    "    vcf_file.write('#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\t')\n",
    "    if(popid_dict == None):\n",
    "        new_ids = []\n",
    "        for i in range(len(geno_dict[0][1])):\n",
    "            curr_nid = ''.join(['ID',str(i)])\n",
    "            new_ids.append(curr_nid)\n",
    "            vcf_file.write('{0}\\t'.format(curr_nid))\n",
    "        vcf_file.write('\\n')\n",
    "    else:\n",
    "        new_ids = []\n",
    "        for i,pop in popid_dict.items():\n",
    "            curr_nid = ''.join(['POP',str(pop),'ID',str(i),])\n",
    "            new_ids.append(curr_nid)\n",
    "            vcf_file.write('{0}\\t'.format(curr_nid))\n",
    "        vcf_file.write('\\n')\n",
    "    for rep,geno_dict in full_geno_dict.items():\n",
    "        is_rep_causalsnp = False\n",
    "        for pos,genos in geno_dict.items():\n",
    "            try:\n",
    "                if(use_diff_chrms == True):\n",
    "                    curr_pos = (int(rep)*(window_spacer+seq_len))+pos\n",
    "                    if(pos == causal_pos_dict[rep]):\n",
    "                        is_rep_causalsnp = True\n",
    "                        vcf_file.write('{0}\\t{1}\\t.\\tA\\tG\\t.\\tPASS\\tCS\\tGT\\t'.format(rep,curr_pos))\n",
    "                    else:\n",
    "                        vcf_file.write('{0}\\t{1}\\t.\\tA\\tG\\t.\\tPASS\\t.\\tGT\\t'.format(rep,curr_pos))\n",
    "                    for g in genos:\n",
    "                        vcf_file.write('{0}|{1}\\t'.format(g[0],g[1]))\n",
    "                    vcf_file.write('\\n')\n",
    "                else:\n",
    "                    curr_pos = (int(rep)*(window_spacer+seq_len))+pos\n",
    "                    if(pos == causal_pos_dict[rep]):\n",
    "                        is_rep_causalsnp = True\n",
    "                        vcf_file.write('1\\t{0}\\t.\\tA\\tG\\t.\\tPASS\\tCS\\tGT\\t'.format(curr_pos))\n",
    "                    else:\n",
    "                        vcf_file.write('1\\t{0}\\t.\\tA\\tG\\t.\\tPASS\\t.\\tGT\\t'.format(curr_pos))\n",
    "                    for g in genos:\n",
    "                        vcf_file.write('{0}|{1}\\t'.format(g[0],g[1]))\n",
    "                    vcf_file.write('\\n')\n",
    "\n",
    "            except:\n",
    "                print(rep,genos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def thetas_toskip(theta):\n",
    "    bad_theta_nums = []\n",
    "    for epoch in range(len(theta)):\n",
    "        curr_theta = theta[:epoch+1]\n",
    "        if(set([all(x == 0) for x in curr_theta[-1][1][1]]) == {True}):\n",
    "            bad_theta_nums.append(epoch)\n",
    "    return bad_theta_nums\n",
    "\n",
    "def init_thetas_toskip(theta):\n",
    "    bad_theta_nums = []\n",
    "    is_init_epoch = True\n",
    "    for epoch in range(len(theta)):\n",
    "        curr_theta = theta[:epoch+1]\n",
    "        if(is_init_epoch == True and set([all(x == 0) for x in curr_theta[-1][1][1]]) == {True}):\n",
    "            bad_theta_nums.append(epoch)\n",
    "        else:\n",
    "            is_init_epoch = False\n",
    "    return bad_theta_nums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Runs the phenotype-generating simulation, for multiple populations and multiple epochs\n",
    "#    theta: list of variables, each entry indicates an epoch. See the description for run_msprime_tskit() to see the full details of this variable\n",
    "#    seq_len (int): length of the sequence that will be simulated\n",
    "#    reps (int): number of replicates to simulate\n",
    "#    pop_schemes (list): list of ints, one per population that will be simulated, indicating the number of individuals that will have 2 genotypes assigned to each\n",
    "#    r (float): recombination rate, provided to the simulate function\n",
    "#    mu (float): mutation rate, provided to the simulate function\n",
    "#    outname (string): name of the file that will be created with the phenotype, one per epoch that is provided, as <outname>.epoch<num_epoch>.phenotypes\n",
    "#    beta (string): method of choosing the beta value that is assigned to each individual. The only kind of distribution that is implemented so far\n",
    "#Returns a 2 lists of dictionaries; per epoch. \n",
    "#    causalpos_byepoch: for each replicate in the epoch, contains the position that was used to calculate the phenotype. Used to indicate in the output vcf file which SNP is the causal one\n",
    "#    genos_byepoch: for each replicate in each epoch, has a dictionary containing the full genotypes for all variants\n",
    "def run_pheno_simulation_multipops(theta,seq_len,reps,pop_schemes,r,mu,outname,beta='normal'):\n",
    "    \n",
    "    genos_byepoch = []\n",
    "    causalpos_byepoch = []\n",
    "    epochs_toskip = thetas_toskip(theta)\n",
    "    for epoch in range(len(theta)):\n",
    "        print('Starting epoch {0} replicates'.format(epoch))\n",
    "\n",
    "        pop_sizes = [(i*2) for i in pop_schemes[epoch]] # Num individuals simulated, equal to twice the number of individuals, since each person is diploid\n",
    "        sample_scheme = [i for i in pop_schemes[epoch]] # Num samples drawn from each pop for likelihood\n",
    "        full_sampsize = sum(sample_scheme)\n",
    "        # print(pop_sizes,sample_scheme)\n",
    "        genotype_index_bypops = assign_genotype_index_multipop(sample_scheme,pop_sizes) #randomly take the <samp_size> number of genomes simulated, and randomly assign to each individual 2 of them\n",
    "        genotype_index_byinds = get_ind_genoindex_multipop(genotype_index_bypops)\n",
    "        # print(genotype_index_byinds)\n",
    "        causalgenotypes_byrep = {x:[] for x in range(reps)}\n",
    "        causalpositions_byrep = {x:0 for x in range(reps)}\n",
    "        fullgenotypes_byrep = {x:[] for x in range(reps)}\n",
    "\n",
    "\n",
    "        curr_theta = theta[:epoch+1]\n",
    "        if(epoch in epochs_toskip):\n",
    "            print('Epoch {0} has no migration, skipping this epoch due to memory constraints'.format(epoch))\n",
    "            continue\n",
    "        # print('epoch {0} theta = {1}'.format(epoch,curr_theta))\n",
    "        tsreps = run_msprime_tskit_v1(curr_theta,pop_sizes,seq_len,r,mu,reps)\n",
    "        # pdb.set_trace()\n",
    "        print('replicate trees created; iterating through genotypes now')\n",
    "\n",
    "        for rep,tree_sequence in enumerate(tsreps): \n",
    "            # pdb.set_trace()\n",
    "            # print('starting replicate {0}'.format(rep))\n",
    "            num_vars = 0\n",
    "            for variant in tree_sequence.variants():\n",
    "                num_vars += 1\n",
    "            causal_var_id = random.randint(0,(num_vars-1))\n",
    "            curr_causal_var = []\n",
    "            curr_causal_pos = 0\n",
    "            curr_full_vars_posgeno_dict = {}\n",
    "            for variant in tree_sequence.variants():\n",
    "                curr_full_vars_posgeno_dict[round(variant.site.position)] = list(variant.genotypes)\n",
    "                if(variant.site.id == causal_var_id):\n",
    "                    curr_causal_var = list(variant.genotypes)\n",
    "                    curr_causal_pos = round(variant.site.position)\n",
    "\n",
    "            if(len(curr_causal_var) != (full_sampsize*2)):\n",
    "                print('not enough causal variants in rep {0}'.format(rep))\n",
    "                curr_causal_var = [0 for x in range((full_sampsize*2))]\n",
    "                curr_full_vars_posgeno_dict[0] = [0 for x in range((full_sampsize*2))]\n",
    "            causalpositions_byrep[rep] = curr_causal_pos\n",
    "\n",
    "            curr_rep_genotypes = []\n",
    "            for indiv,index in genotype_index_byinds.items():\n",
    "                try:\n",
    "                    curr_rep_genotypes.append((curr_causal_var[index[0]],curr_causal_var[index[1]]))\n",
    "                except:\n",
    "                    print(indiv,index,len(curr_causal_var))\n",
    "            causalgenotypes_byrep[rep] = [curr_causal_pos,curr_rep_genotypes]\n",
    "\n",
    "            curr_rep_fullgenos = {}\n",
    "            for pos,geno in curr_full_vars_posgeno_dict.items():\n",
    "                temp_fullgeno = []\n",
    "                for indiv,index in genotype_index_byinds.items():\n",
    "                    try:\n",
    "                        temp_fullgeno.append((geno[index[0]],geno[index[1]]))\n",
    "                    except:\n",
    "                        print(rep,indiv,index,len(geno))\n",
    "                curr_rep_fullgenos[pos] = temp_fullgeno\n",
    "            fullgenotypes_byrep[rep] = curr_rep_fullgenos\n",
    "\n",
    "        print('finished iterating through trees, starting the phenotype calculations')\n",
    "        causalgenotypes_byind = {x:[] for x in range(full_sampsize)}\n",
    "        for pos,genos in causalgenotypes_byrep.items():\n",
    "            for num,g in enumerate(genos[1]):\n",
    "                causalgenotypes_byind[num].append(g)\n",
    "\n",
    "        beta = 'normal'\n",
    "        phenotypes_byinds = {x:0 for x in range(full_sampsize)}\n",
    "        beta_list = generate_betas(num_inds=reps,dist_type=beta)\n",
    "        for i in range(full_sampsize):\n",
    "            # curr_beta = beta_list[i]\n",
    "            phenotypes_byinds[i] = estimate_pheno(causalgenotypes_byind[i],beta_list)\n",
    "        # phenos_byepoch.append(phenotypes_byinds)\n",
    "#         write_phenofile('{0}.epoch{1}'.format(outname,epoch),phenotypes_byinds)\n",
    "        \n",
    "        genos_byepoch.append(fullgenotypes_byrep)\n",
    "        causalpos_byepoch.append(causalpositions_byrep)\n",
    "    \n",
    "    return causalpos_byepoch,genos_byepoch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_input_file_full_v1(filename):\n",
    "    infile = open(filename,'r')\n",
    "    theta_true = []\n",
    "    sample_scheme_byepoch = []\n",
    "    curr_epoch = -1\n",
    "    migmats_byepoch = {}\n",
    "    times_byepoch = []\n",
    "    curr_epochtype = ''\n",
    "    massmigs_byepoch = {}\n",
    "    demoevs_byepoch = {}\n",
    "    ooafrica_sampscheme = []\n",
    "    for line in infile:\n",
    "        if(line[0] == '#'):\n",
    "            curr_epoch += 1\n",
    "            cline = line.split('\\n')[0][1:].split('\\t')\n",
    "            curr_epochtype = cline[0]\n",
    "            if(curr_epochtype == 'ooafrica'):\n",
    "                curr_popscheme = cline[1]\n",
    "                curr_epoch -= 1\n",
    "                ooafrica_sampscheme = [int(x) for x in curr_popscheme.split(',')]    # Num of samples per population in the observed data\n",
    "                # sample_scheme_byepoch.append(ooafrica_sampscheme)\n",
    "            else:\n",
    "                curr_time = cline[1]\n",
    "                times_byepoch.append(float(curr_time))\n",
    "                curr_popscheme = cline[2]\n",
    "                pop_sizes = [int(x) for x in curr_popscheme.split(',')]    # Num of samples per population in the observed data\n",
    "                sample_scheme_byepoch.append(pop_sizes)\n",
    "        else:\n",
    "            curr_mm = [float(f) for f in line.split('\\n')[0].split('\\t')]\n",
    "            if(curr_epoch not in demoevs_byepoch.keys()):\n",
    "                demoevs_byepoch[curr_epoch] = [curr_epochtype]\n",
    "            demoevs_byepoch[curr_epoch].append(curr_mm)\n",
    "\n",
    "#     if(args.ooafrica):\n",
    "    \n",
    "    # pop sizes\n",
    "    N_base = 12300    # Number of diploids in the population; based on the N_AF from the tutorial Out_Of_Africa model\n",
    "    for e in range(len(sample_scheme_byepoch)):\n",
    "        Ne0 = np.ones(len(sample_scheme_byepoch[e]))*N_base\n",
    "        theta_true.append([demoevs_byepoch[e][0],times_byepoch[e],[Ne0,np.array(demoevs_byepoch[e][1:])]])\n",
    "    if(len(ooafrica_sampscheme) > 0):\n",
    "        theta_true.insert(0,['ooafrica',ooafrica_sampscheme,[]])\n",
    "        sample_scheme_byepoch.insert(0,ooafrica_sampscheme)\n",
    "    return theta_true,sample_scheme_byepoch\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_theta,test_sampsch = read_input_file_full_v1('/local3/jake/admix_simul/testing/single_massmig_event/input_file.migr_events.v2')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "920.0\n"
     ]
    }
   ],
   "source": [
    "ooa_pop_configs, ooa_mig, ooa_demoevents = outofafrica_model_parameters(test_sampsch[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mass migration: Lineages moved with probability 1.0 backwards in time with source 2 & dest 1\n",
      "                     (equivalent to migration from 1 to 2 forwards in time)\n",
      "Migration rate change to 0 everywhere\n",
      "Migration rate change for (0, 1) to 0.00015\n",
      "Migration rate change for (1, 0) to 0.00015\n",
      "Population parameter change for 1: initial_size -> 1861 growth_rate -> 0 \n",
      "Mass migration: Lineages moved with probability 1.0 backwards in time with source 1 & dest 0\n",
      "                     (equivalent to migration from 0 to 1 forwards in time)\n",
      "Population parameter change for 0: initial_size -> 7300 \n"
     ]
    }
   ],
   "source": [
    "for d in ooa_demoevents:\n",
    "    print(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 864,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 0, 1: 0, 2: 0, 3: 0, 4: 1, 5: 1, 6: 1, 7: 1, 8: 2, 9: 2, 10: 2, 11: 2}"
      ]
     },
     "execution_count": 864,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "popindex_byind = {x:0 for x in range(full_sampsize)}\n",
    "for pop,size in enumerate(sample_scheme):\n",
    "    for x in range(size):\n",
    "        popindex_byind[(x + (size*pop))] = pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{5: [(0, 1),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (1, 0),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (1, 1)],\n",
       " 184: [(0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (1, 1)],\n",
       " 244: [(0, 1),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (1, 1)],\n",
       " 340: [(0, 0),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0)],\n",
       " 357: [(0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0)],\n",
       " 464: [(0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (1, 0),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0)],\n",
       " 539: [(0, 0),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0)],\n",
       " 547: [(0, 0),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0)],\n",
       " 568: [(0, 1),\n",
       "  (0, 0),\n",
       "  (1, 0),\n",
       "  (1, 0),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (1, 1)],\n",
       " 800: [(1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (1, 0),\n",
       "  (0, 0)],\n",
       " 855: [(1, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (1, 0),\n",
       "  (0, 0)],\n",
       " 857: [(0, 1),\n",
       "  (1, 1),\n",
       "  (1, 1),\n",
       "  (1, 1),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (0, 1),\n",
       "  (1, 1),\n",
       "  (1, 0),\n",
       "  (0, 0),\n",
       "  (0, 1),\n",
       "  (1, 1)]}"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting epoch 0 replicates\n",
      "0 8 4 [0, 1, 2, 3, 4, 5, 6, 7]\n",
      "1 8 4 [8, 9, 10, 11, 12, 13, 14, 15]\n",
      "2 8 4 [16, 17, 18, 19, 20, 21, 22, 23]\n",
      "[[0.     0.0001 0.01  ]\n",
      " [0.0001 0.     0.001 ]\n",
      " [0.01   0.001  0.    ]]\n",
      "there is only one epoch, starting now using theta [[0.0, [array([10000., 10000., 10000.]), array([[0.    , 0.0001, 0.01  ],\n",
      "       [0.0001, 0.    , 0.001 ],\n",
      "       [0.01  , 0.001 , 0.    ]])]]]\n",
      "replicate trees created; iterating through genotypes now\n",
      "finished iterating through trees, starting the phenotype calculations\n",
      "Starting epoch 1 replicates\n",
      "0 8 4 [0, 1, 2, 3, 4, 5, 6, 7]\n",
      "1 14 7 [8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21]\n",
      "2 14 7 [22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35]\n",
      "Epoch 1 has no migration, skipping this epoch due to memory constraints\n",
      "Starting epoch 2 replicates\n",
      "0 10 5 [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]\n",
      "1 12 6 [10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21]\n",
      "2 14 7 [22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35]\n",
      "[[0.     0.0001 0.01  ]\n",
      " [0.0001 0.     0.001 ]\n",
      " [0.01   0.001  0.    ]]\n",
      "replicate trees created; iterating through genotypes now\n",
      "finished iterating through trees, starting the phenotype calculations\n",
      "Starting epoch 3 replicates\n",
      "0 8 4 [0, 1, 2, 3, 4, 5, 6, 7]\n",
      "1 6 3 [8, 9, 10, 11, 12, 13]\n",
      "2 4 2 [14, 15, 16, 17]\n",
      "[[0.     0.0001 0.01  ]\n",
      " [0.0001 0.     0.001 ]\n",
      " [0.01   0.001  0.    ]]\n",
      "replicate trees created; iterating through genotypes now\n",
      "finished iterating through trees, starting the phenotype calculations\n"
     ]
    }
   ],
   "source": [
    "# write_genovcf(g[0],c[0],'/local3/jake/admix_simul/testing/test_multipop_geno',full_sampsize,seq_len=1000,popid_dict=popindex_byind)\n",
    "test_theta = read_input_file_full('/local3/jake/admix_simul/testing/test_inputfile.migration_matrix')\n",
    "c,g = run_pheno_simulation_multipops(test_theta[0],1000,5,test_theta[1],2e-8,2e-8,'/local3/jake/admix_simul/testing/test_multipop_pheno.v3')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "m_AF_AS = 3e-5 #migration rate of african to non-african\n",
    "m_EU_AS = 9.6e-5 #migration rate of non-african to african\n",
    "\n",
    "time_migration = 3000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "2\n",
      "3\n",
      "4\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes_byinds = {x:0 for x in range(full_sampsize)}\n",
    "beta_list = generate_betas(num_inds=reps,dist_type=beta)\n",
    "for i in range(full_sampsize):\n",
    "    curr_beta = beta_list[i]\n",
    "    phenotypes_byinds[i] = estimate_pheno(causalgenotypes_byind[i],curr_beta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_sample_draws(draws_max,pop_assign,sample_scheme):\n",
    "    \"\"\"\n",
    "    Compute draws of the observed samples to use for likelihood calculation\n",
    "    \"\"\"\n",
    "    draws = []\n",
    "    \n",
    "    n = np.sum(sample_scheme)\n",
    "    Np = len(sample_scheme)\n",
    "    draws = []\n",
    "    while (len(draws) < draws_max):\n",
    "        draw = []\n",
    "        for i in range(Np):\n",
    "            ni = sample_scheme[i]\n",
    "            samples_i = sorted(np.random.choice(pop_assign[i],size=ni,replace=False))\n",
    "            draw.extend(samples_i)\n",
    "            \n",
    "        draw_tup = tuple(draw)\n",
    "        if draw_tup not in draws:\n",
    "            draws.append(draw_tup)\n",
    "    return draws\n",
    "\n",
    "def simulate_sequence_data(theta,sample_nums,L,r,mu):\n",
    "    # Does not incorporate recombination maps. \n",
    "    S = []  # To return an L x N array, where N=sum(sample_nums)\n",
    "    \n",
    "    N = np.sum(sample_nums)\n",
    "    print('Target number of sites in the seq:',L)    \n",
    "    S = np.zeros((L,N),dtype=int)\n",
    "    \n",
    "    ts_replicates = run_msprime_tskit(theta,sample_nums,L,r,mu,1)   # There is only 1 ARG in the replicates iterator\n",
    "    \n",
    "    for tree_seq in ts_replicates:\n",
    "        Nvar = 0\n",
    "        Nfix_der = 0\n",
    "        Nfix_anc = 0\n",
    "        \n",
    "        x_last = 0\n",
    "        for variant in tree_seq.variants():\n",
    "            x = int(variant.site.position)\n",
    "            \n",
    "            d = x-x_last\n",
    "            nanc = np.max([d-1,0])   \n",
    "            Nfix_anc += nanc\n",
    "            \n",
    "            S[x,:] = list(variant.genotypes)\n",
    "            nd = np.sum(S[x,:])\n",
    "            if nd == 0:\n",
    "                Nfix_anc += 1\n",
    "            else:\n",
    "                Nvar += 1\n",
    "                if nd == N:\n",
    "                    Nfix_der += 1\n",
    "            x_last = x\n",
    "\n",
    "    print('Nvar:',Nvar)\n",
    "    print('Nfix_der',Nfix_der)\n",
    "    print('Nfix_anc:',Nfix_anc)\n",
    "    print('Ntot:',Nvar+Nfix_anc)\n",
    "    \n",
    "    return S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.7976931348623156e+300"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1.7976931348623157e+300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "920.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([<msprime.simulations.PopulationConfiguration at 0x7f746606acc0>,\n",
       "  <msprime.simulations.PopulationConfiguration at 0x7f746606a080>,\n",
       "  <msprime.simulations.PopulationConfiguration at 0x7f746606afd0>],\n",
       " [[0, 2.5e-05, 7.8e-06], [2.5e-05, 0, 3.11e-05], [7.8e-06, 3.11e-05, 0]],\n",
       " [{'type': 'mass_migration', 'time': 920.0, 'source': 2, 'dest': 1, 'proportion': 1.0},\n",
       "  {'type': 'migration_rate_change', 'time': 920.0, 'rate': 0, 'matrix_index': None},\n",
       "  {'type': 'migration_rate_change', 'time': 920.0, 'rate': 0.00015, 'matrix_index': (0, 1)},\n",
       "  {'type': 'migration_rate_change', 'time': 920.0, 'rate': 0.00015, 'matrix_index': (1, 0)},\n",
       "  {'type': 'population_parameters_change', 'time': 920.0, 'growth_rate': 0, 'initial_size': 1861, 'population': 1},\n",
       "  {'type': 'mass_migration', 'time': 2040.0, 'source': 1, 'dest': 0, 'proportion': 1.0},\n",
       "  {'type': 'population_parameters_change', 'time': 5920.0, 'growth_rate': None, 'initial_size': 7300, 'population': 0}])"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outofafrica_model_parameters([100,0,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#This function sets up the out_of_africa model, outlined in the msprime tutorial and expanded by user \"slowkoni\" on github, from the Gravel 2011 paper\n",
    "#these values will be used to initialize a simulation so that additional user-definied events can simulated in more recent times\n",
    "def outofafrica_model_parameters(n_samples):\n",
    "    generation_time = 25\n",
    "\n",
    "#     n_samples = [10,10,10]\n",
    "\n",
    "    m_AF_B = 15e-5\n",
    "    m_AF_EU = 2.5e-5\n",
    "    m_AF_AS = 0.78e-5\n",
    "    m_EU_AS = 3.11e-5\n",
    "\n",
    "    N_A = 7300 #ancestral_size \n",
    "    N_AF = 14474 #africa_size \n",
    "    N_EU0 = 1032 #out-to-europe-size\n",
    "    N_AS0 = 550 #out-to-asia-size\n",
    "    N_AF_B = 1861 #out-of-africa-size\n",
    "\n",
    "    T_B = 51000 #merge-to-africa time\n",
    "    T_EuAs = 23000 #merge-europe-asia-time\n",
    "    T_AF = 148000 #africa-expansion-time\n",
    "\n",
    "    asia_growth_rate = 0.38 \n",
    "    europe_growth_rate = 0.48\n",
    "\n",
    "     # Calculate the final sizes, or in coalescent terms, the starting sizes, of the\n",
    "    # Europe and East Asia populations which under the Gravel et al. 2011 out of africa\n",
    "    # model experience exponential growth after diverging from each other going\n",
    "    # separate ways in the world. Command line accepts values in percentages as given in\n",
    "    # the Gravel et al. paper. We just need to know the starting effective size, starting\n",
    "    # as in the present day effective size, as the simulation goes backward in time.\n",
    "    europe_final_size = N_EU0 / math.exp(-(europe_growth_rate/100.) * (T_EuAs/generation_time))\n",
    "    asia_final_size = N_AS0 / math.exp(-(asia_growth_rate/100.) * (T_EuAs/generation_time))\n",
    "\n",
    "    migration_matrix = [\n",
    "        [      0, m_AF_EU, m_AF_AS],\n",
    "        [m_AF_EU,       0, m_EU_AS],\n",
    "        [m_AF_AS, m_EU_AS,       0],\n",
    "    ]\n",
    "    \n",
    "    europe_asia_merge_time = T_EuAs/generation_time\n",
    "    print(europe_asia_merge_time)\n",
    "    population_configurations = [\n",
    "        ms.PopulationConfiguration(\n",
    "            sample_size=n_samples[0], initial_size=N_AF),\n",
    "        ms.PopulationConfiguration(\n",
    "            sample_size=n_samples[1], initial_size=europe_final_size,\n",
    "            growth_rate=europe_growth_rate/100.),\n",
    "        ms.PopulationConfiguration(\n",
    "            sample_size=n_samples[2], initial_size=asia_final_size,\n",
    "            growth_rate=asia_growth_rate/100.)\n",
    "    ]\n",
    "\n",
    "    demographic_events = [\n",
    "        # All the next events, until indicated below, are coincident and thus really one\n",
    "        # event\n",
    "        \n",
    "        # The Europe (pop 1) and East Asia (pop 2) populations merge and become population 1\n",
    "        ms.MassMigration(\n",
    "            time=europe_asia_merge_time, source=2, destination=1, proportion=1.0),\n",
    "        # Migration rates must now refelect migration between the original African population\n",
    "        # and the merged Europe/East Asian population which are still separate at this point\n",
    "        ms.MigrationRateChange(time=europe_asia_merge_time, rate=0),\n",
    "        ms.MigrationRateChange(\n",
    "            time=europe_asia_merge_time, rate=m_AF_B, matrix_index=(0, 1)),\n",
    "        ms.MigrationRateChange(\n",
    "            time=europe_asia_merge_time, rate=m_AF_B, matrix_index=(1, 0)),\n",
    "        # Finally, the effective size of the merged population (pop 1 now) is different\n",
    "        # otherwise it would be whatever the shrinking Europe population size is/was.\n",
    "        ms.PopulationParametersChange(\n",
    "            time=europe_asia_merge_time, initial_size=N_AF_B,\n",
    "            growth_rate=0, population_id=1),\n",
    "\n",
    "        # Now we are backward further in time\n",
    "        # Next, the migrating out of africa population (currently pop 1) joins to the\n",
    "        # population of origin (pop 0), the african population\n",
    "        ms.MassMigration(\n",
    "            time=T_B/generation_time, source=1, destination=0, proportion=1.0),\n",
    "\n",
    "        # In the final event of the model, the African population (pop 0, the only one left)\n",
    "        # reduces in size to the ancestral population size until the MRCA (end of the simulation)\n",
    "        ms.PopulationParametersChange(\n",
    "            time=T_AF/generation_time,\n",
    "            initial_size=N_A, population_id=0)\n",
    "    ]\n",
    "\n",
    "    return(population_configurations, migration_matrix, demographic_events)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[4, 4, 4]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_scheme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def run_msprime_tskit_v1(theta,sample_nums,L,r,mu,R,outname=None,outname_epoch=None):\n",
    "    ts_replicates = None\n",
    "    \n",
    "    Np = len(sample_nums)\n",
    "#     print('samples to be drawn for populations: {0}'.format(sample_nums))\n",
    "    # if(args.ooafrica):\n",
    "    #     init_pop_configs, init_mig, init_demoevents = outofafrica_model_parameters()\n",
    "    demo_events = []\n",
    "\n",
    "    if(theta[0][0] == 'ooafrica'):\n",
    "        init_pop_configs, init_mig, demo_events = outofafrica_model_parameters(theta[0][1])\n",
    "    else:\n",
    "        init_pop_configs = [ms.PopulationConfiguration(sample_size=sample_nums[i], initial_size=theta[0][2][0][i]) for i in range(Np)]\n",
    "        \n",
    "        if(theta[0][0] == 'mrate'):\n",
    "            init_mig = theta[0][2][1]\n",
    "        elif(theta[0][0] == 'mass'):\n",
    "            init_mig = None\n",
    "            for m in theta[0][2][1]:\n",
    "                demo_events.append(ms.MassMigration(time=theta[0][1], source=m[0], destination=m[1], proportion=m[2]))\n",
    "    # if(args.ooafrica):\n",
    "    #     ooa_pop_configs, ooa_mig, ooa_demoevents = outofafrica_model_parameters()\n",
    "    #     for d in ooa_demoevents:\n",
    "    #         init_demoevents.append(d)\n",
    "    K = len(theta)  # K = number of epochs\n",
    "    print(theta)\n",
    "    if K > 1:\n",
    "        # There is more than one epoch, so must set the non-initial epochs as demographic events\n",
    "        # demo_events = []\n",
    "        for k in range(1,K):\n",
    "            theta_type,t_k,theta_k = theta[k]\n",
    "            \n",
    "\n",
    "            if(theta_type == 'mrate'):\n",
    "                Ne = theta_k[0]\n",
    "                mig = theta_k[1]\n",
    "                for i in range(Np):\n",
    "                    # Set the Ne\n",
    "                    demo_events.append(ms.PopulationParametersChange(population=i,time=t_k,initial_size=Ne[i]))\n",
    "                    print('matrix to add: {0}'.format(mig))\n",
    "\n",
    "                    # Set the migration rates\n",
    "                    for j in range(Np):\n",
    "                        if j!=i:\n",
    "                            demo_events.append(ms.MigrationRateChange(time=t_k,rate=mig[i,j],matrix_index=tuple([i,j])))\n",
    "            elif(theta_type == 'mass'):\n",
    "                Ne = theta_k[0]\n",
    "                massmig = theta_k[1]\n",
    "                for m in massmig:\n",
    "                    demo_events.append(ms.MassMigration(time=t_k, source=m[0], destination=m[1], proportion=m[2]))\n",
    "            elif(theta_type == 'ooafrica'):\n",
    "                ooa_pop_configs, ooa_mig, ooa_demoevents = outofafrica_model_parameters(theta[k][1])\n",
    "                for d in ooa_demoevents:\n",
    "                    demo_events.append(d)\n",
    "            else:\n",
    "                print('Epoch events should only be \"mass\" or \"mrate\"!')\n",
    "                return -1\n",
    "        # if(args.ooafrica):\n",
    "        #     ooa_pop_configs, ooa_mig, ooa_demoevents = outofafrica_model_parameters()\n",
    "        #     for d in ooa_demoevents:\n",
    "        #         demo_events.append(d)\n",
    "#         print(init_pop_configs)\n",
    "        ts_replicates = ms.simulate(\n",
    "            length=L,\n",
    "            recombination_rate=r,\n",
    "            population_configurations=init_pop_configs,\n",
    "            migration_matrix = init_mig,\n",
    "            demographic_events=demo_events,\n",
    "            num_replicates=R,\n",
    "            mutation_rate = mu\n",
    "        )\n",
    "    else:\n",
    "        print('there is only one epoch, starting now using theta {0}'.format(theta))\n",
    "        # if(args.ooafrica):\n",
    "        #     ooa_pop_configs, ooa_mig, ooa_demoevents = outofafrica_model_parameters()\n",
    "        #     for d in ooa_demoevents:\n",
    "        #         print(d)\n",
    "        #         init_demoevents.append(d)\n",
    "        #     if(init_mig is None): #The only event provided is a mass migration event, and the user specifies the out-of-africa flag, we can use that migration matrix as the initial migration matrix\n",
    "        #         init_mig = ooa_mig \n",
    "        # There is only the initial epoch. \n",
    "        if(init_mig is None): #If the first epoch event is a mass migration event, and the out-of-africa is not provided, then there's no migration matrix to provide\n",
    "            ts_replicates = ms.simulate(\n",
    "            length=L,\n",
    "            recombination_rate=r,\n",
    "            population_configurations=init_pop_configs,\n",
    "            demographic_events = demo_events,\n",
    "            num_replicates=R,\n",
    "            mutation_rate = mu)\n",
    "        else:\n",
    "            if(len(demo_events) > 0):\n",
    "                ts_replicates = ms.simulate(\n",
    "                length=L,\n",
    "                recombination_rate=r,\n",
    "                population_configurations=init_pop_configs,\n",
    "                demographic_events = demo_events,\n",
    "                migration_matrix = init_mig,\n",
    "                num_replicates=R,\n",
    "                mutation_rate = mu)\n",
    "            else:\n",
    "                ts_replicates = ms.simulate(\n",
    "                    length=L,\n",
    "                    recombination_rate=r,\n",
    "                    population_configurations=init_pop_configs,\n",
    "                    migration_matrix = init_mig,\n",
    "                    num_replicates=R,\n",
    "                    mutation_rate = mu\n",
    "                )\n",
    "    if(demo_events != None):\n",
    "        dp = ms.DemographyDebugger(\n",
    "            population_configurations=init_pop_configs,\n",
    "            migration_matrix=init_mig,\n",
    "            demographic_events=demo_events)\n",
    "        if(outname == None):\n",
    "            print('Demography History for epoch {0}'.format(outname_epoch))\n",
    "            dp.print_history(output=sys.stderr)\n",
    "        else:\n",
    "            outfile = open('{0}.epoch{1}.demohistory'.format(outname,outname_epoch),'w')\n",
    "            dp.print_history(output=outfile)\n",
    "            outfile.close()\n",
    "    return ts_replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ID0', 'ID1', 'ID2', 'ID3', 'ID4', 'ID5', 'ID6', 'ID7', 'ID8', 'ID9']"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def out_of_africa():\n",
    "    # First we set out the maximum likelihood values of the various parameters\n",
    "    # given in Table 1.\n",
    "    N_A = 7300\n",
    "    N_B = 2100\n",
    "    N_AF = 12300\n",
    "    N_EU0 = 1000\n",
    "    N_AS0 = 510\n",
    "    # Times are provided in years, so we convert into generations.\n",
    "    generation_time = 25\n",
    "    T_AF = 220e3 / generation_time\n",
    "    T_B = 140e3 / generation_time\n",
    "    T_EU_AS = 21.2e3 / generation_time\n",
    "    # We need to work out the starting (diploid) population sizes based on\n",
    "    # the growth rates provided for these two populations\n",
    "    r_EU = 0.004\n",
    "    r_AS = 0.0055\n",
    "    N_EU = N_EU0 / math.exp(-r_EU * T_EU_AS)\n",
    "    N_AS = N_AS0 / math.exp(-r_AS * T_EU_AS)\n",
    "    # Migration rates during the various epochs.\n",
    "    m_AF_B = 25e-5\n",
    "    m_AF_EU = 3e-5\n",
    "    m_AF_AS = 1.9e-5\n",
    "    m_EU_AS = 9.6e-5\n",
    "    # Population IDs correspond to their indexes in the population\n",
    "    # configuration array. Therefore, we have 0=YRI, 1=CEU and 2=CHB\n",
    "    # initially.\n",
    "    population_configurations = [\n",
    "        msprime.PopulationConfiguration(\n",
    "            sample_size=0, initial_size=N_AF),\n",
    "        msprime.PopulationConfiguration(\n",
    "            sample_size=1, initial_size=N_EU, growth_rate=r_EU),\n",
    "        msprime.PopulationConfiguration(\n",
    "            sample_size=1, initial_size=N_AS, growth_rate=r_AS)\n",
    "    ]\n",
    "    migration_matrix = [\n",
    "        [      0, m_AF_EU, m_AF_AS],\n",
    "        [m_AF_EU,       0, m_EU_AS],\n",
    "        [m_AF_AS, m_EU_AS,       0],\n",
    "    ]\n",
    "    demographic_events = [\n",
    "        # CEU and CHB merge into B with rate changes at T_EU_AS\n",
    "        msprime.MassMigration(\n",
    "            time=T_EU_AS, source=2, destination=1, proportion=1.0),\n",
    "        msprime.MigrationRateChange(time=T_EU_AS, rate=0),\n",
    "        msprime.MigrationRateChange(\n",
    "            time=T_EU_AS, rate=m_AF_B, matrix_index=(0, 1)),\n",
    "        msprime.MigrationRateChange(\n",
    "            time=T_EU_AS, rate=m_AF_B, matrix_index=(1, 0)),\n",
    "        msprime.PopulationParametersChange(\n",
    "            time=T_EU_AS, initial_size=N_B, growth_rate=0, population_id=1),\n",
    "        # Population B merges into YRI at T_B\n",
    "        msprime.MassMigration(\n",
    "            time=T_B, source=1, destination=0, proportion=1.0),\n",
    "        # Size changes to N_A at T_AF\n",
    "        msprime.PopulationParametersChange(\n",
    "            time=T_AF, initial_size=N_A, population_id=0)\n",
    "    ]\n",
    "    # Use the demography debugger to print out the demographic history\n",
    "    # that we have just described.\n",
    "    dd = msprime.DemographyDebugger(\n",
    "        population_configurations=population_configurations,\n",
    "        migration_matrix=migration_matrix,\n",
    "        demographic_events=demographic_events)\n",
    "#     dd.print_history()\n",
    "#     tree_sequence = msprime.simulate(sample_size=samp_size, Ne=1e5, length=seq_len, recombination_rate=2e-8,mutation_rate=2e-8, \n",
    "#                                  random_seed=10)\n",
    "#     tree_sequence = msprime.simulate(population_configurations=population_configurations,demographic_events=demographic_events,\n",
    "#                                      Ne=N_AF, length=seq_len, recombination_rate=2e-8,mutation_rate=2e-8,migration_matrix=migration_matrix)\n",
    "    return dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = out_of_africa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'DemographyDebugger' object cannot be interpreted as an integer",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-158-cb41d8a9c60e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mmsprime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimulate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/msprime-env/lib/python3.7/site-packages/msprime/simulations.py\u001b[0m in \u001b[0;36msimulate\u001b[0;34m(sample_size, Ne, length, recombination_rate, recombination_map, mutation_rate, population_configurations, migration_matrix, demographic_events, samples, model, record_migrations, random_seed, mutation_generator, num_replicates, from_ts, start_time, end_time, record_full_arg, num_labels)\u001b[0m\n\u001b[1;32m    445\u001b[0m         \u001b[0mstart_time\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstart_time\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    446\u001b[0m         \u001b[0mrecord_full_arg\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrecord_full_arg\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 447\u001b[0;31m         num_labels=num_labels)\n\u001b[0m\u001b[1;32m    448\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    449\u001b[0m     parameters = {\n",
      "\u001b[0;32m~/anaconda3/envs/msprime-env/lib/python3.7/site-packages/msprime/simulations.py\u001b[0m in \u001b[0;36msimulator_factory\u001b[0;34m(sample_size, Ne, random_generator, length, recombination_rate, recombination_map, population_configurations, migration_matrix, samples, demographic_events, model, record_migrations, from_ts, start_time, end_time, record_full_arg, num_labels)\u001b[0m\n\u001b[1;32m    208\u001b[0m                 \"simultaneously.\")\n\u001b[1;32m    209\u001b[0m         \u001b[0ms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSample\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpopulation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 210\u001b[0;31m         \u001b[0mthe_samples\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0ms\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0m_\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_size\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    211\u001b[0m     \u001b[0;31m# If we have population configurations we may have embedded sample_size\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    212\u001b[0m     \u001b[0;31m# values telling us how many samples to take from each population.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'DemographyDebugger' object cannot be interpreted as an integer"
     ]
    }
   ],
   "source": [
    "msprime.simulate(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
